### ISSUE TITLE
gost locking up after about 500 sequential HTTP inserts

### ISSUE THREAD

**Issue by justb4** [Geodan/gost#85]
I waited long to submit this issue, but can't find the cause and apart from optimizing our publisher I am out of ideas. Hopefully this may look familiar.

In our setup we do not directly publish from sensors to the STA. Reason a.o. is that the server performs refinement steps like validation, calibration and aggregation (hour values) putting all data in a PostGIS table. From this table GeoServer serves WMS-(Time) and WFS. Dedicated "publisher" ETL processes (via cron) publish the latest refined data from PostGIS to STA (and SOS via SOS-T). Sync-tracking is used to determine the latest inserts. When the SOS/STA is empty bulk inserts may be performed (we now have around 2.5 million obs). 

This went well for the SOS (and SensorUp STA) but with GOST I notice the following:

* after about 500 HTTP requests GOST locks up and does not accept new HTTP connections
* Dashboard still responds but shows empty pages (maybe internal locks not released?)
* GOST then needs to be restarted and also Dashboard works again (until next lockup)  
* happens no matter which host/path we use: external URL, internal Docker IP or linked gost Docker Container name
* also happens when we use smaller batches: e.g. 100 records (then at about 3rd batch)
* lockup remains also after Publisher process has long exited
* no error messages in `docker logs --follow gost`

The HTTP requests are "complete", i.e. no reuse of HTTP-connections yet, e.g. each Observation in single HTTP request (need optimization also with STA Batch/Array). Also GET requests are performed, e.g. on first Observation to see if Thing/DataStream etc exists, or Location has changed (PATCH), so a mixture of all types of HTTP requests.

The `STAPublisher` that performs inserts is a specialized [Stetl](http://stetl.org) Output component:
https://github.com/Geonovum/smartemission/blob/master/etl/staoutput.py. 

It could be that the client does not nicely close all HTTP connections, but the publisher process does not run continuously (is scheduled with cron each 10 mins).

**Comment by bertt** [comment]
Ok next week we can work on benchmark test (https://github.com/Geodan/gost/issues/55) for inserts, see if we can reproduce this issue.
To be sure: the test is sending a batch of http post to /Observations, is that correct?

**Comment by bertt** [comment]
Partly related issue: https://github.com/Geodan/gost/issues/11

**Comment by justb4** [comment]
New finding: could never trigger lockup when publishing client is on other machine from gost server, e.g. from home...All lockups seen were when both publishing client and gost both run as Docker containers within the same server instance. Docker does some clever things with connections...

**Comment by bertt** [comment]
some questions: can you reproduce this lockup thing with running your JMeter-Docker as client (in same Docker server instance)? Only with POST or also with GET's?

**Comment by tebben** [comment]
I was unable to reproduce the lock with JMeter but was able to get a lock while spamming the send button in postman on "Post thing with location and datastream" ill try to fix this and hope it also fixes your issue.

**JMeter Test**
150.000 GET requests to /v1.0/Datastreams?$top=1&$select=name,description&$expand=Observations($select=result;$top=5)
150.000 POST requests on /v1.0/Datastreams(1)/Observations

-------------------------------
JMeter: Windows
GOST(Latest): Ubuntu development machine
Database(Latest): Docker
GET - OK
POST - OK
-------------------------------
JMeter: Windows
GOST(Latest): Docker
Database(Latest): Docker
GET - OK
POST - OK
-------------------------------
JMeter: Docker
GOST(Latest): Docker
Database(Latest): Docker
GET - OK
POST - OK 
-------------------------------

**Comment by tebben** [comment]
Locking after creating Things with location was caused by an unclosed connection on inserting the HistoricalLocation which resulted in a full connection pool. It looks like this was the only place the connection was not closed properly, spamming post Thing with location from postman causes no locks anymore. 

Can you tell me what you are posting when the lock happens?

**Comment by justb4** [comment]
The overal behavior looks like a connection pool being exhausted. I first thought it was the DB connection pool as at first I could not connect to our PostGIS Docker container, also not from other apps. I increased the PostGIS connection pool in `postgresql.conf` but still the lockup happened, but the PostGIS container was reachable. The other observation (see above) was that the Dashboard could still be loaded but all Entities were empty. After restart of `gost` they were visible. 

But it could be the http connection pool as well as you observed. It is hard for me to determine when the lockup happened. I tried a `curl` script that POSTs forever `Thing` but never a lockup happened.

Our `stapublisher` uses a Python class [staoutput.py](https://github.com/Geonovum/smartemission/blob/master/etl/staoutput.py) that receives a stream (records) of measurements and intelligently (by querying `gost` and building an internal STA model), determines what to POST/PATCH also from sensor metadata:

*  when no supporting Entity (`Thing`, `Datastream`,...) for the Observation is present: POST these Entities first
* `Location`: is POSTed before `Thing` and referenced in Thing-post-data (so no embedded `Location` in `Thing`)
* `Location`: when `Thing` present in `gost`, `Location` is PATCHed with new coords
* finally POST `Observation`

So the sequence of POSTed data is dependent what is already present in `gost`. At some point only `Observations` are POSTed (as all supporting Entities will be present) and Locations PATCHed. This way new sensors can be added without explicit management of `gost` Entities and no prior knowledge of Entity id's is required in the client.

I have attached the output `stapublisher.txt` that happens on first run with empty below. After about 496 POST requests (no PATCHes, as all Things are first-timers) and the 20th `Thing` lockup happens. Always at the same point in the stream: the last POST is Thing number 20. The next GET then blocks. 

Logfile (13355 lines!)
[stapublisher.txt](https://github.com/Geodan/gost/files/928681/stapublisher.txt)

**Comment by justb4** [comment]
Forgot to mention: above was on `gost` Docker 0.4 and `latest` before your commit 1e06f83 above, will try that now via Docker `latest`....

**Comment by justb4** [comment]
And good news: no lockup found after commit 1e06f83 !! Thanks and closing...

**Comment by justb4** [comment]
Sorry, but the lockup still occurs, it just takes longer, plus after a connection timeout (exactly 5 mins IMO the standard TCP timeout) POSTing will resume until the next lock. Happens both with direct publication to `gost` Container or via the Apache proxy Container. See the log output below. Three `Observations` are POSTed, first is ok, second has timeout, and directly after the third is ok again. Since the response comes from Apache, I can see that it comes from the backend `gost` Container. The Dashboard is still accessible and shows entities...May need to rule out if this lockup is Docker-specific.

```
2017-04-18 14:25:24,351 staoutput INFO ====START POST Observation id=12-co-38900
2017-04-18 14:25:24,351 staoutput INFO POST to URL: http://test.smartemission.nl:80/gost/v1.0/Observations ...
2017-04-18 14:25:24,351 urllib3.connectionpool INFO Starting new HTTP connection (1): test.smartemission.nl
2017-04-18 14:25:24,356 staoutput INFO Req nr 39445 - response status: code=201 msg=201
2017-04-18 14:25:24,356 staoutput INFO Response Headers: CaseInsensitiveDict({'content-length': '559', 'server': 'Apache/2.4.7 (Ubuntu)', 'location': 'http://test.smartemission.nl/gost/v1\
.0/Observations(38900)', 'date': 'Tue, 18 Apr 2017 12:25:24 GMT', 'access-control-allow-origin': '*', 'content-type': 'application/json; charset=UTF-8'})
2017-04-18 14:25:24,356 staoutput INFO YES added Observation! id=12-co-38900 status=201
2017-04-18 14:25:24,357 staoutput INFO ====END POST Observation id=12-co-38900
2017-04-18 14:25:24,357 staoutput INFO ====START POST Observation id=12-temperature-38901
2017-04-18 14:25:24,357 staoutput INFO POST to URL: http://test.smartemission.nl:80/gost/v1.0/Observations ...
2017-04-18 14:25:24,357 urllib3.connectionpool INFO Starting new HTTP connection (1): test.smartemission.nl
2017-04-18 14:30:24,374 staoutput INFO Req nr 39446 - response status: code=502 msg=502
2017-04-18 14:30:24,374 staoutput INFO Response Headers: CaseInsensitiveDict({'date': 'Tue, 18 Apr 2017 12:25:24 GMT', 'content-length': '510', 'content-type': 'text/html; charset=iso-885\
9-1', 'server': 'Apache/2.4.7 (Ubuntu)'})
2017-04-18 14:30:24,374 staoutput WARNING FAIL POST Observation status=502 payload={
  "Datastream": {
    "@iot.id": 425
  },
  "phenomenonTime": "2016-06-16T10:00:00+00:00",
  "result": 19,
  "resultTime": "2016-06-16T10:00:00+00:00",
  "parameters": {
      "gid": 38901, "raw_gid": 11529, "station": 12, "name": "temperature"
  }
}
 res=<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>502 Proxy Error</title>
</head><body>
<h1>Proxy Error</h1>
<p>The proxy server received an invalid^M
response from an upstream server.<br />^M
The proxy server could not handle the request <em><a href="/gost/v1.0/Observations">POST&nbsp;/gost/v1.0/Observations</a></em>.<p>
Reason: <strong>Error reading from remote server</strong></p></p>
<hr>
<address>Apache/2.4.7 (Ubuntu) Server at test.smartemission.nl Port 80</address>
</body></html>

2017-04-18 14:30:24,374 staoutput INFO ====END POST Observation id=12-temperature-38901
2017-04-18 14:30:24,374 staoutput INFO ====START POST Observation id=12-coraw-38902
2017-04-18 14:30:24,374 staoutput INFO POST to URL: http://test.smartemission.nl:80/gost/v1.0/Observations ...
2017-04-18 14:30:24,375 urllib3.connectionpool INFO Starting new HTTP connection (1): test.smartemission.nl
2017-04-18 14:30:24,380 staoutput INFO Req nr 39447 - response status: code=201 msg=201
2017-04-18 14:30:24,380 staoutput INFO Response Headers: CaseInsensitiveDict({'content-length': '563', 'server': 'Apache/2.4.7 (Ubuntu)', 'location': 'http://test.smartemission.nl/gost/v1\
.0/Observations(38902)', 'date': 'Tue, 18 Apr 2017 12:30:24 GMT', 'access-control-allow-origin': '*', 'content-type': 'application/json; charset=UTF-8'})
2017-04-18 14:30:24,380 staoutput INFO YES added Observation! id=12-coraw-38902 status=201
2017-04-18 14:30:24,380 staoutput INFO ====END POST Observation id=12-coraw-38902
```

**Comment by tebben** [comment]
Test posting 300.000 Observations using JMeter

gost:latest, gost-db:latest, Apache as reverse proxy
result: 502 errors

gost:latest, gost-db:latest
result: Never returning requests (lost in space)

gost:latest, local database (docker mdillon/postgis:9.5), Apache
result: no problems found

It looks like Apache has some timeouts set by default returning a 502 when a request takes too long. Because of no detected problems with my local database I changed gost-db to use mdillon/postgis:9.6 instead of mdillon/postgis:9.5-alpine this seems like to fix the problem, after sending 300.000 Observations using 60 threads no problems were detected.

gost-db:latest now uses postgis 9.6 which cannot use the old 9.5 files which will load by default if you ran gost before so be sure to remove the old postgis volume from the machine first.

**Comment by justb4** [comment]
Interesting. So the problem seems the DB connections in combination with Docker, not the Golang HTTP pool. Or at least when both gost and PG are Dockerized. I can add to this:

* both with Apache reverse proxy and direct gost-latest connection we get the lockup problem
* smaller batches e.g. 1000 inserts per client process invocation does not help, there is a memory effect
* initially I increased the number of PG connections to 200 in postgresql.conf, that helped, but just took longer before lockup. With the default value (think 120) the DB locked up (refused connections) also for all other client-processes.

But now the interesting part: as we already had a PostGIS Docker Container we don't use `gost-db`. We simply (re)create the gost schema in our existing PG DB mapped to local file volume. We use [kartoza/postgis](https://github.com/kartoza/docker-postgis) version 9.4-2.1 based on `debian:stable`, see our [Dockerfile](https://github.com/Geonovum/smartemission/blob/master/docker/postgis/Dockerfile) (just for setting timezone right).

So we use both a lower PG version (9.4 PostGIS 2.1) and different Docker Image. Have you tried  `mdillon/postgis:9.6-alpine`? If that works it could be a PG version issue. Kartoza does not yet have a 9.6 version. If it is really a PG version issue I can look into upgrading. But looks we're getting further. Thanks!

**Comment by justb4** [comment]
Hmm. Reimported both  `gost:latest` (that included commit 380cac511721265bc3c667f3299d3bd5824df5d4) and `gost-db:latest` (verified it had PG 9.6) but still can trigger lockup by mass POSTing similar as with our original PostGIS Docker Image `kartoza/postgis version 9.4-2.1` setup. Sometimes the client can POST 20000 Entities in one run, but mostly the lockup occurs, not after specific amount of POSTs, can be 500 or 17000 or any other number, seems random. Observations:

* When stopping the POSTing client when locked, I can re-run the client. It is then not immediately locked, but can start POST until blocked after random number of POSTs
* In all cases the external API URL `/gost/v1.0` remains functioning, even if client blocked. 

I ran `tcpflow` on the `gost` Docker container (`tcpflow -i docker0 -c host 172.17.0.7`) to see the data passing from the client (`172.17.0.10`) to the STA API and from `gost` (`172.17.0.7`) to `gost-db` (`172.17.0.9`).

The normal flow when POSTing an `Observation` is for example:

```
172.017.000.010.44727-172.017.000.007.08080: POST /v1.0/Observations HTTP/1.1
Content-Length: 252
Accept-Encoding: gzip, deflate, compress
Accept: */*
User-Agent: Stetl Python http
Host: gost
Content-Type: application/json;charset=UTF-8

{
  "Datastream": {
    "@iot.id": 618
  },
  "phenomenonTime": "2016-04-18T23:00:00+00:00",
  "result": 1021,
  "resultTime": "2016-04-18T23:00:00+00:00",
  "parameters": {
      "gid": 67618, "raw_gid": 14874, "station": 13, "name": "pressure"
  }
}


172.017.000.007.46172-172.017.000.009.05432: QnSELECT location.id AS A_location_id, location.name AS A_location_name, location.description AS A_location_description, location.encodingtype AS A_location_encodingtype, public.ST_AsGeoJSON(location.location) AS A_location_location FROM v1.location INNER JOIN LATERAL (SELECT datastream.id AS datastream_id FROM v1.datastream INNER JOIN v1.thing_to_location ON location.id = thing_to_location.location_id AND thing_to_location.thing_id = datastream.thing_id ) AS datastream on true  WHERE location.id IN (SELECT location.id FROM v1.location INNER JOIN LATERAL (SELECT datastream.id AS datastream_id FROM v1.datastream INNER JOIN v1.thing_to_location ON location.id = thing_to_location.location_id AND thing_to_location.thing_id = datastream.thing_id ) AS datastream on true  WHERE datastream.datastream_id = 618 ORDER BY location.id DESC  OFFSET 0) ORDER BY location.id DESC


172.017.000.009.05432-172.017.000.007.46172: Ta_location_idEa_location_nameEa_location_descriptionEa_location_encodingtypeEa_location_locationDh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.84


172.017.000.009.05432-172.017.000.007.46172: 1171]}CSELECT 12ZI


172.017.000.007.46172-172.017.000.009.05432: QFselect id from v1.featureofinterest where original_location_id=64


SELECT 1ZI0.009.05432-172.017.000.007.46172: TidED64C


172.017.000.007.46172-172.017.000.009.05432: QINSERT INTO v1.observation (data, stream_id, featureofinterest_id) VALUES ('{"phenomenonTime":"2016-04-18T23:00:00.000Z","result":1021,"resultTime":"2016-04-18T23:00:00+00:00","parameters":{"gid":67618,"name":"pressure","raw_gid":14874,"station":13}}', 618, 64) RETURNING id


172.017.000.009.05432-172.017.000.007.46172: TidFFD126303CINSERT 0 1ZI


172.017.000.007.08080-172.017.000.010.44727: HTTP/1.1 201 Created
Access-Control-Allow-Origin: *
Content-Type: application/json; charset=UTF-8
Location: http://test.smartemission.nl/gost/v1.0/Observations(126303)
Date: Mon, 24 Apr 2017 11:13:47 GMT
Content-Length: 570

{
   "@iot.id": 126303,
   "@iot.selfLink": "http://test.smartemission.nl/gost/v1.0/Observations(126303)",
   "phenomenonTime": "2016-04-18T23:00:00.000Z",
   "result": 1021,
   "parameters": {
      "gid": 67618,
      "name": "pressure",
      "raw_gid": 14874,
      "station": 13
   },
   "Datastream@iot.navigationLink": "http://test.smartemission.nl/gost/v1.0/Observations(126303)/Datastream",
   "FeatureOfInterest@iot.navigationLink": "http://test.smartemission.nl/gost/v1.0/Observations(126303)/FeatureOfInterest",
   "resultTime": "2016-04-18T23:00:00+00:00"
}

```

But when a lockup occurs the dataflow is as below: The INSERT is successful but the POST response is never returned. So it looks like locking happens after the INSERT and before returning the POST response, at least from the `tcpflow` capture.

```
172.017.000.010.44727-172.017.000.007.08080: POST /v1.0/Observations HTTP/1.1
Content-Length: 254
Accept-Encoding: gzip, deflate, compress
Accept: */*
User-Agent: Stetl Python http
Host: gost
Content-Type: application/json;charset=UTF-8

{
  "Datastream": {
    "@iot.id": 619
  },
  "phenomenonTime": "2016-04-18T23:00:00+00:00",
  "result": 2,
  "resultTime": "2016-04-18T23:00:00+00:00",
  "parameters": {
      "gid": 67619, "raw_gid": 14874, "station": 13, "name": "noiselevelavg"
  }
}


172.017.000.007.46172-172.017.000.009.05432: QnSELECT location.id AS A_location_id, location.name AS A_location_name, location.description AS A_location_description, location.encodingtype AS A_location_encodingtype, public.ST_AsGeoJSON(location.location) AS A_location_location FROM v1.location INNER JOIN LATERAL (SELECT datastream.id AS datastream_id FROM v1.datastream INNER JOIN v1.thing_to_location ON location.id = thing_to_location.location_id AND thing_to_location.thing_id = datastream.thing_id ) AS datastream on true  WHERE location.id IN (SELECT location.id FROM v1.location INNER JOIN LATERAL (SELECT datastream.id AS datastream_id FROM v1.datastream INNER JOIN v1.thing_to_location ON location.id = thing_to_location.location_id AND thing_to_location.thing_id = datastream.thing_id ) AS datastream on true  WHERE datastream.datastream_id = 619 ORDER BY location.id DESC  OFFSET 0) ORDER BY location.id DESC


172.017.000.009.05432-172.017.000.007.46172: Ta_location_idEa_location_nameEa_location_descriptionEa_location_encodingtypeEa_location_locationDh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.841171]}Dh6413Location of Station 1313{"type":"Point","coordinates":[5.838721,51.84


172.017.000.009.05432-172.017.000.007.46172: 1171]}CSELECT 12ZI


172.017.000.007.46172-172.017.000.009.05432: QFselect id from v1.featureofinterest where original_location_id=64


SELECT 1ZI0.009.05432-172.017.000.007.46172: TidED64C


172.017.000.007.46172-172.017.000.009.05432: QINSERT INTO v1.observation (data, stream_id, featureofinterest_id) VALUES ('{"phenomenonTime":"2016-04-18T23:00:00.000Z","result":2,"resultTime":"2016-04-18T23:00:00+00:00","parameters":{"gid":67619,"name":"noiselevelavg","raw_gid":14874,"station":13}}', 619, 64) RETURNING id


172.017.000.009.05432-172.017.000.007.46172: TidFFD126304CINSERT 0 1ZI

```

**Comment by justb4** [comment]
Ok, looks like it is the absence of an MQTT service will cause lockup under strain. Our setup now includes the Mosquitto MQTT Docker container and since running not a single lockup has been seen!

Possibly with the other lockup causes/fixes this issue looks resolved for now. When stress-testing a lot always comes out!

Still our config had disabled MQTT, so that is something for issue #88. Our `config.yaml` had:

```
mqtt:
    enabled: false
    host: localhost
    port: 1883
```
It looks like [the code](https://github.com/Geodan/gost/blob/master/src/sensorthings/api/observation.go#L163) does not reflect this. At least I think `gost` should be able to run without MQTT but I am glad we can close this issue, hopefully for good!
